packages/rpc-methods/src/permitted/requestSnaps.ts

```diff
@@ -0,0 +1,158 @@
+import { ethErrors } from 'eth-rpc-errors';
+import {
+  PermissionConstraint,
+  RequestedPermissions,
+} from '@metamask/controllers';
+import {
+  PermittedHandlerExport,
+  JsonRpcRequest,
+  PendingJsonRpcResponse,
+  JsonRpcEngineEndCallback,
+} from '@metamask/types';
+import { getSnapPermissionName } from '@metamask/snap-utils';
+import { hasProperty, isObject } from '@metamask/utils';
+import {
+  handleInstallSnaps,
+  InstallSnapsHook,
+  InstallSnapsResult,
+} from './common/snapInstallation';
+
+/**
+ * `wallet_requestSnaps` installs the requested Snaps and requests permission to use them if necessary.
+ */
+export const requestSnapsHandler: PermittedHandlerExport<
+  RequestSnapsHooks,
+  RequestedPermissions,
+  InstallSnapsResult
+> = {
+  methodNames: ['wallet_requestSnaps'],
+  implementation: requestSnapsImplementation,
+  hookNames: {
+    installSnaps: true,
+    requestPermissions: true,
+    getPermissions: true,
+  },
+};
+
+export type RequestSnapsHooks = {
+  /**
+   * Installs the requested snaps if they are permitted.
+   */
+  installSnaps: InstallSnapsHook;
+
+  /**
+   * Initiates a permission request for the requesting origin.
+   *
+   * @returns The result of the permissions request.
+   */
+  requestPermissions: (
+    permissions: RequestedPermissions,
+  ) => Promise<PermissionConstraint[]>;
+
+  /**
+   * Gets the current permissions for the requesting origin.
+   *
+   * @returns The current permissions of the requesting origin.
+   */
+  getPermissions: () => Promise<
+    Record<string, PermissionConstraint> | undefined
+  >;
+};
+
+/**
+ * Checks whether existing permissions satisfy the requested permissions
+ *
+ * Note: Currently, we don't compare caveats, if any caveats are requested, we always return false.
```

Comment:
Why?

Reply:
We don't have a good way to compare caveats for equality, so we expect the permissions we already have to not be enough if caveats are involved.

This won't be a problem in reality for now, because the permissions in question will only be snap permissions that don't have caveats.